### DISCUSSION TITLE
Inking performance: Prediction algorithm? Settings? OS?

### DISCUSSION THREAD

**Discussion by Bini613** [xournalpp/xournalpp#3954] [Q&A]
Does Xounral++'s stylus/pen processing include some sort of prediction algorithm? I really like this app and the community behind it, but the main thing holding me back is the inking performance. Microsoft inking has gotten so responsive in recent times, and from what I understand, its prediction algorithm is largely to thank for this. In Microsoft inking apps, on my Samsung galaxy book pro 360, there is no perceivable latency between the pen tip and the ink trail - no matter how fast I move it. Xournal++ as far as I've experienced, is good, but can't compare. I've tried tinkering with the input settings, but I don't really understand them, and I seem only to be capable of making it worse. So I have a few questions:

Is there a settings configuration that can bring Xournal++ closer to Microsoft Inking's performance? 
Is there there a prediction algorithm in Xournal++, and will it be improved or implemented?
Will I see better performance if I move to Linux on my system, instead of Windows (11)?

Thanks

**Comment by rolandlo** [comment]
There is no prediction algorithm in Xournal++. Instead you can try out the input stabilization in the input system panel in menu `Edit > Preferences`, although that has a slightly different purpose, and may not bring you closer to MS Inking's performance. Currently no one is working on a prediction algorithm and your suggestion is new to us.

In my experience, Xournal++'s performance is best on Linux, but since Xournal++ version 1.1.1 it is close to that on Windows. On my X1 Yoga I cannot observe any latency.

**Reply by Bini613** [reply]
Thanks for explaining. What are the chances of a prediction algorithm coming in the future? I have some knowledge on the math that would be needed, (e.g the Runge–Kutta approximation method) and it seems fairly simple computationally - though I don't know how easy it would be to implement in C++. I've heard it needs 'low level api access' or something like that, to be implemented with justifiably low latency. Can someone who understands this stuff comment?

**Reply by personalizedrefrigerator** [reply]
See also https://github.com/xournalpp/xournalpp/pull/2729

**Reply by flxzt** [reply]
I came across this [stroke prediction library](https://google.github.io/ink-stroke-modeler/#implementation-details) while doing some research for my own app. Might be worth taking a look at :)

**Reply by LittleHuba** [reply]
Thanks for mentioning this library @flxzt !
We should definitely check if we can drop this into the input system of Xournal++.

**Reply by 0skxr** [reply]
Has anything moved in this area in the last 2 years?

**Reply by rolandlo** [reply]
@0skxr No, I wouldn't know of anyone who has worked on it.

**Comment by BLACK4585** [comment]
Hey,
I also tried to increase the performance of my pen with it's input delay. I'm running a Surface Pro 4 with Nobara, which has the features of linux-kernel already patched in the kernel. I installed the other tools provided by kernel-surface like IPTDS (or how it's called) and some other stuff to make the whole touch experience working.
I tried playing around with the stabilization settings and the best was Gaussian weighting with deadzone. But this still doesn't provide the same performance as e.g. on RNote, which has notable better input delay. Also in Linux ofc.
Maybe someone who knows this stuff could look into this, what they do different. Although this is written in Rust, maybe there are some things regarding the input which can be transferred over to X++?
Also, did someone look into this stroke prediction library yet?
